zero_bin/rpc/src/jumpdest.rs

```diff
@@ -0,0 +1,471 @@
+use alloy::providers::ext::DebugApi;
+use alloy::providers::Provider;
+use alloy::rpc::types::eth::Transaction;
+use alloy::rpc::types::trace::geth::StructLog;
+use alloy::rpc::types::trace::geth::{GethDebugTracingOptions, GethDefaultTracingOptions};
+use alloy::transports::RpcError;
+use alloy::transports::Transport;
+use alloy::transports::TransportErrorKind;
+use alloy_primitives::B256;
+use alloy_primitives::U256;
+use anyhow::ensure;
+use evm_arithmetization::jumpdest::JumpDestTableWitness;
+use keccak_hash::keccak;
+use structlogprime::normalize_structlog;
+use trace_decoder::TxnTrace;
+use tracing::trace;
+
+/// Pass `true` for the components needed.
+fn structlog_tracing_options(stack: bool, memory: bool, storage: bool) -> GethDebugTracingOptions {
+    GethDebugTracingOptions {
+        config: GethDefaultTracingOptions {
+            disable_stack: Some(!stack),
+            // needed for CREATE2
+            disable_memory: Some(!memory),
+            disable_storage: Some(!storage),
+            ..GethDefaultTracingOptions::default()
+        },
+        tracer: None,
+        ..GethDebugTracingOptions::default()
+    }
+}
+
+fn trace_contains_create2(structlog: Vec<StructLog>) -> bool {
+    structlog.iter().any(|entry| entry.op == "CREATE2")
+}
+
+// Gets the lightest possible structlog for transcation `tx_hash`.
+pub(crate) async fn get_normalized_structlog<ProviderT, TransportT>(
+    provider: &ProviderT,
+    tx_hash: &B256,
+) -> Result<Option<Vec<StructLog>>, RpcError<TransportErrorKind>>
+where
+    ProviderT: Provider<TransportT>,
+    TransportT: Transport + Clone,
+{
+    let light_structlog_trace = provider
+        .debug_trace_transaction(*tx_hash, structlog_tracing_options(false, false, false))
+        .await?;
+
+    let structlogs_opt: Option<Vec<StructLog>> = normalize_structlog(light_structlog_trace).await;
+
+    let need_memory = structlogs_opt.is_some_and(trace_contains_create2);
+
+    let structlog = provider
+        .debug_trace_transaction(
+            *tx_hash,
+            structlog_tracing_options(true, need_memory, false),
+        )
+        .await?;
+
+    let ret = normalize_structlog(structlog).await;
+    Ok(ret)
+}
+
+/// Provides a way to check in constant time if an address points to a
+/// precompile.
+fn precompiles() -> &'static HashSet<Address> {
+    static PRECOMPILES: OnceLock<HashSet<Address>> = OnceLock::new();
+    PRECOMPILES.get_or_init(|| {
+        HashSet::<Address>::from_iter((1..=0xa).map(|x| Address::from(U160::from(x))))
+    })
+}
+
+/// Generate at JUMPDEST table by simulating the call stack in EVM,
+/// using a Geth structlog as input.
+pub(crate) fn generate_jumpdest_table(
+    tx: &Transaction,
+    struct_log: &[StructLog],
+    tx_traces: &BTreeMap<Address, TxnTrace>,
+) -> anyhow::Result<JumpDestTableWitness> {
+    trace!("Generating JUMPDEST table for tx: {}", tx.hash);
+    ensure!(struct_log.is_empty().not(), "Structlog is empty.");
```

Comment:
This is still present in the code, hence failing to preprocess JDTs on any batch containing asset transfers

Reply:
gone now, thanks.